timemachine/fe/mcgregor.py

```diff
@@ -303,242 +396,161 @@ def mcs(
-                recursion(
-                    g1,
-                    g2,
-                    atom_map_1_to_2,
-                    atom_map_2_to_1,
-                    layer + 1,
-                    new_marcs,
-                    mcs_result,
-                    priority_idxs,
-                    max_visits,
-                    max_cores,
-                    threshold,
-                    enforce_core_core,
-                    max_connected_components,
-                    min_connected_component_size,
-                    filter_fxn,
-                    leaf_filter_fxn,
-                )
-            atom_map_pop(atom_map_1_to_2, atom_map_2_to_1, layer, jdx)
-
-    # always allow for explicitly not mapping layer atom
-    # nit: don't need to check for connected core if mapping to None
-    new_marcs = refine_marcs(g1, g2, layer, UNMAPPED, marcs)
-
-    recursion(
-        g1,
-        g2,
-        atom_map_1_to_2,
-        atom_map_2_to_1,
-        layer + 1,
-        new_marcs,
-        mcs_result,
-        priority_idxs,
-        max_visits,
-        max_cores,
-        threshold,
-        enforce_core_core,
-        max_connected_components,
-        min_connected_component_size,
-        filter_fxn,
-        leaf_filter_fxn,
-    )
+) -> Callable[[Node, int], Tuple[Sequence[Node], int]]:
+    def satisfies_connected_components_constraints(node: Node) -> bool:
+        if max_connected_components is not None or min_connected_component_size > 1:
+            g1_mapped_nodes = {a1 for a1, a2 in enumerate(node.atom_map.a_to_b[: node.layer]) if a2 != UNMAPPED}
+
+            if g1_mapped_nodes:
+                # Nodes in g1 are visited in order, so nodes left to visit are [layer, layer + 1, ..., n - 1]
+                g1_unvisited_nodes = set(range(node.layer, g1.n_vertices))
+                if g1.mapping_incompatible_with_cc_constraints(
+                    g1_mapped_nodes, g1_unvisited_nodes, max_connected_components, min_connected_component_size
+                ):
+                    return False
+
+            g2_mapped_nodes = {a2 for a2, a1 in enumerate(node.atom_map.b_to_a) if a1 != UNMAPPED}
+
+            if g2_mapped_nodes:
+                # Nodes in g2 are visited in the order determined by priority_idxs. Nodes may be repeated in priority_idxs,
+                # but we skip over nodes that have already been mapped.
+                g2_unvisited_nodes = {
+                    a2 for a2s in priority_idxs[node.layer :] for a2 in a2s if a2 not in g2_mapped_nodes
+                }
+                if g2.mapping_incompatible_with_cc_constraints(
+                    g2_mapped_nodes, g2_unvisited_nodes, max_connected_components, min_connected_component_size
+                ):
+                    return False
+
+        return True
+
+    def expand(node: Node, best_num_edges: int) -> Tuple[List[Node], int]:
+        if node.marcs.num_edges_upper_bound < best_num_edges:
```

Comment:
does it make sense to also check `node.marcs.num_edges_upper_bound < min_num_edges`? i.e. return if the predicate is `True`

Reply:
Since `best_num_edges` is initialized to `min_num_edges`, `min_num_edges <= best_num_edges` is an invariant, so I think this would be redundant.